File: Lab/Lab_1_2/src/Lab_2/MergeSort.py
from src.Lab_1.ReadFile import ReadFile


def merge(arr, start, mid, end):
    if len(arr) == 0:
        return arr

    lenArrL = mid - start + 1
    lenArrR = end - mid

    arrL = arr[start:mid + 1]
    arrR = arr[mid + 1:end + 1]

    i = 0
    j = 0
    k = start

    while i < lenArrL and j < lenArrR:
        if arrL[i] < arrR[j]:
            arr[k] = arrL[i]
            k = k + 1
            i = i + 1
        else:
            arr[k] = arrR[j]
            k = k + 1
            j = j + 1

    while i < lenArrL:
        arr[k] = arrL[i]
        k = k + 1
        i = i + 1
    while j < lenArrR:
        arr[k] = arrR[j]
        k = k + 1
        j = j + 1

    return arr


def sort(arr, start, end):
    if (start < end):
        mid = start + (end - start) // 2
        sort(arr, start, mid)
        sort(arr, mid + 1, end)
        merge(arr, start, mid, end)
    return arr


array = ReadFile.first_main()

print(array)
start = 0;
end = len(array) - 1
mid = start + (end - start) // 2
print("Sorted: ")
print(sort(array, start, end))

File: Lab/Lab_1_1/src/main/java/ip/group/Lab_2/MergeSort.java
package ip.group.Lab_2;

import ip.group.Lab_1.ReadFile;

import java.util.Arrays;

import static ip.group.Lab_1.ReadFile.fistMain;

public class MergeSort {

    public static int[] sort(int[] arr, int start, int end) {
        if (start < end) {

            int mid = start + (end - start) / 2;
            sort(arr, start, mid);
            sort(arr, mid + 1, end);
            merge(arr, start, mid, end);
        }
        return arr;
    }

    public static int[] merge(int[] arr, int start, int mid, int end) {
        if (arr.length == 0)
            return arr;

        int arrL = mid - start + 1;
        int arrR = end - mid;
        int k = start;

        int[] arr1 = new int[arrL];
        int[] arr2 = new int[arrR];

        for (int i = 0; i < arrL; i++)
            arr1[i] = arr[start + i];
        for (int j = 0; j < arrR; j++)
            arr2[j] = arr[mid + 1 + j];

        int i = 0, j = 0;
        while (i < arrL && j < arrR)
            if (arr1[i] <= arr2[j])
                arr[k++] = arr1[i++];
            else
                arr[k++] = arr2[j++];

        while (i < arrL)
            arr[k++] = arr1[i++];
        while (j < arrR)
            arr[k++] = arr2[j++];

        return arr;
    }

    public static void main(String[] args) {
        //Scanner scan = new Scanner(System.in);
        //String strArr = scan.nextLine();

        ReadFile readFile = new ReadFile();

        int[] arrayInt = readFile.toNumberSplit(fistMain());
        System.out.println(Arrays.toString(sort(arrayInt, 0, arrayInt.length - 1)));
    }
}
